New UI

tanrax opened the pull request:
Would that be interesting?

tanrax committed:
Added new UI

tanrax committed:
Updated documentation

tanrax committed:
Moved comments button

basil-conto commented on hackernews.el:
```diff
@@ -326,6 +339,108 @@ This is intended as an :annotation-function in
+      (when background-color
+        (push `(:background ,background-color) props))
+      (when props
+        (put-text-property start end 'face (apply #'append props))))))
```
It would be more configurable and play better with Font Lock if we used named faces defined via `defface` rather than ad-hoc anonymous faces. Ideally the default colours would adapt well to the user's background mode etc.

basil-conto commented on hackernews.el:
```diff
@@ -326,6 +339,108 @@ This is intended as an :annotation-function in
+                 :notify (lambda (&rest _)
+                           (hackernews-top-stories))
+                 :help-echo "View top stories"
+                 " 🔥 Top ")
```
I don't think we can assume that emoji will always be displayable in the user's system/font, so this should degrade gracefully or at least be configurable. IIRC more recent Emacs versions introduced a mechanism for this, namely `define-icon` or similar: [`(info "(elisp) Icons")`](https://gnu.org/s/emacs/manual/html_node/elisp/Icons.html)

basil-conto commented on hackernews.el:
```diff
@@ -345,40 +460,31 @@ This is intended as an :annotation-function in
+  (interactive)
+  (let ((separator-regex (concat "^" (regexp-quote (hackernews--string-separator)) "$")))
+    (if (search-forward-regexp separator-regex nil t)
+        (forward-line 2)
```
My first reaction is that this is too tightly coupled with the display format. I would prefer to avoid search-based motion, instead relying on either some feature of the UI, or some marker (e.g. text property) that we place in the buffer that is independent of how the buffer is displayed.

basil-conto commented on hackernews.el:
```diff
@@ -532,61 +638,108 @@ This is for compatibility with various Emacs versions.
-         (inhibit-read-only t))
+         (feed  (hackernews--get :feed))
+         (feed-name (hackernews--feed-name feed))
+         (is-first-load (= (point-max) 1)))
```
In general there's no need to compare against concrete position values in Elisp. If you want to check whether the narrowed portion of the buffer is empty, you can compare `point-min` to `point-max`; and for checking whether the buffer is empty irrespective of narrowing, there is `buffer-size`.

basil-conto commented on hackernews.el:
```diff
@@ -532,61 +638,108 @@ This is for compatibility with various Emacs versions.
+         (is-first-load (= (point-max) 1)))
+
+    ;; Temporarily disable read-only mode
+    (read-only-mode -1)
```
Why not bind `inhibit-read-only`?

basil-conto commented on hackernews.el:
```diff
@@ -532,61 +638,108 @@ This is for compatibility with various Emacs versions.
 (defun hackernews--display-items ()
   "Render items associated with, and pop to, the current buffer."
+  ;; Switch to buffer first (like Lobsters)
+  (switch-to-buffer (current-buffer))
```
Why is this needed? In general `switch-to-buffer` is a user-facing command, and Elisp programs should be able to get away with `with-current-buffer`, `pop-to-buffer`, etc. instead.

basil-conto commented on hackernews.el:
```diff
@@ -532,61 +638,108 @@ This is for compatibility with various Emacs versions.
 
+    ;; Set up the buffer with hackernews-mode (after inserting content)
+    (when is-first-load
+      (hackernews-mode))
```
What is the significance of enabling the major mode after inserting the buffer contents?

basil-conto commented on hackernews.el:
```diff
@@ -532,61 +638,108 @@ This is for compatibility with various Emacs versions.
+    ;; Setup widgets
+    (widget-setup)
+
+    ;; Disable line numbers
+    (when (fboundp 'display-line-numbers-mode)
+      (display-line-numbers-mode 0))
```
Why is this needed? Ideally we shouldn't have to override user preferences.

basil-conto commented on hackernews.el:
```diff
@@ -624,7 +777,16 @@ Official major mode key bindings:
+             (require 'visual-fill-column nil t))
+    (setq-local visual-fill-column-center-text t)
+    (setq-local visual-fill-column-width hackernews-display-width)
+    (visual-fill-column-mode 1))
```
Why does `hackernews` need to be aware of `visual-fill-column-mode`, and load and enable it explicitly? Can't the user enable the minor mode as desired via `hackernews-mode-hook`? If the setup is nontrivial we can always document it in our README/wiki.

basil-conto commented on hackernews.el:
```diff
@@ -624,7 +777,16 @@ Official major mode key bindings:
+    (setq-local visual-fill-column-width hackernews-display-width)
+    (visual-fill-column-mode 1))
+  ;; Activate the keymap
+  (use-local-map hackernews-mode-map))
```
Why is this needed? Usually this is done automatically by `define-derived-mode`.

basil-conto commented on hackernews.el:
```diff
@@ -698,9 +860,16 @@ rendered at the end of the hackernews buffer."
     (with-current-buffer (get-buffer-create (format "*hackernews %s*" name))
       (unless append
+        ;; Kill local variables BEFORE disabling read-only (like Lobsters)
+        (kill-all-local-variables)
```
Why is this needed?

basil-conto commented on hackernews.el:
```diff
@@ -698,9 +860,16 @@ rendered at the end of the hackernews buffer."
+        (read-only-mode -1)
+
+        ;; Clear buffer
         (let ((inhibit-read-only t))
```
Why is this needed given the subsequent `inhibit-read-only`?

tanrax replied:
technically it is not necessary

basil-conto reviewed (commented):
Thank you very much for the contribution! There is definitely interest in improving the UI.

I won't have time to review this carefully for a little while, so here are some thoughts after a quick scan of the changes.

- The topic of improving the UI has been brought up before. Have you looked over #44, #45, and #55? Since those discussions, I have also learned of another attractive UI/MVC framework that is built into Emacs, namely `ewoc`: [`(info "(elisp) Abstract Display")`](https://gnu.org/s/emacs/manual/html_node/elisp/Abstract-Display.html)

- Ideally the choice of UI should be configurable via a user option. Similarly, we should try to avoid changing default key bindings (and their effects) if possible.

basil-conto commented on hackernews.el:
```diff
@@ -1,11 +1,12 @@
 ;; Keywords: comm hypermedia news
-;; Version: 0.7.1
+;; Version: 0.8.0
+;; Package-Requires: ((emacs "24.3"))
```
Either:
- remove the versions lower than 24.3 from `.github/workflows/build.yml`; or
- add a dependency on the `cl-lib` compatibility package from GNU ELPA: https://elpa.gnu.org/packages/cl-lib.html

Personally I am happy to raise the minimum required Emacs version as needed, given that the `hackernews` package talks to remote third-party machines over the network.

tanrax commented:
Thank you very much for the detailed review and feedback!

 Previous discussions and ewoc:  I'll review those now. The ewoc framework sounds interesting; I wasn't familiar with it. 
Configurable UI: The default could remain the current minimal UI, and users could opt into the enhanced
  version. Does this sound like the right approach?

Regarding specific code issues:

- Anonymous faces vs defface: You're completely right. I'll refactor this.
- Emoji support: Good point. It should be configurable.
- Search-based navigation: I see the issue with coupling navigation to the visual format. Would you recommend using text properties or markers to identify item boundaries?
- Position comparisons: I'll change (= (point-max) 1) to (= (buffer-size) 0) or (= (point-min) (point-max)).
- The inhibit-read-only, switch-to-buffer, mode timing, and other issues: These were carryovers from my attempts to get the code working. I will make changes.

Well...would you prefer that I:

  1. Make all these fixes in the current PR, or
  2. Close this PR and open a new one after studying ewoc and the previous discussions more carefully?

basil-conto commented:
> Configurable UI: The default could remain the current minimal UI, and users could opt into the enhanced version. Does this sound like the right approach?

That would be amazing for backward compatibility, which though a general goal of the Emacs project, may or may not be our highest priority for the `hackernews` project. However, a secondary benefit of designing the new UI to be optional is that the resulting solution will hopefully be general enough to afford any number of future UIs :).

If it proves too much of a hassle we can always reconsider.

>     * Search-based navigation: I see the issue with coupling navigation to the visual format. Would you recommend using text properties or markers to identify item boundaries?

Markers can degrade Emacs redisplay performance (at least this was the case historically in Emacs, though more recent changes to Emacs internals may have mitigated this). The penalty may be negligible in the case of `hackernews`, but if all else is the same, then my vote is for plain text properties.

However, if the Widget/EWOC/etc. library provides functions for locating UI elements, then I think we should prefer that over a custom solution.

> Well...would you prefer that I:
> 
>     1. Make all these fixes in the current PR, or
> 
>     2. Close this PR and open a new one after studying ewoc and the previous discussions more carefully?

It's completely up to you, AFAIC there's no harm in leaving this PR open and in progress.

Thanks again, and happy hacking! :)

tanrax committed:
Added optional emojis

tanrax committed:
Add configurable UI styles (classic/modern)

tanrax commented:
This commit implements a dual UI system, allowing users to choose between the classic minimalist interface (default) and the new interface. I believe it doesn't break any compatibility issues.

It is configured with the `hackernews-ui-style` variable with values ​​'classic (default)' and 'modern'. For example:

```
(use-package hackernews
    :ensure t
    :config
        (setq hackernews-ui-style 'modern)
       (setq hackernews-enable-emojis t))

(global-set-key (kbd "C-c x h") 'hackernews)
```

tanrax committed:
Add visual-fill-column for centered modern UI

tanrax committed:
Set version to 0.8.0

tanrax committed:
Fix modern UI: use 'face instead of 'font-lock-face and fullscreen display

tanrax committed:
Fix modern UI navigation and widgets interaction

tanrax committed:
Add configuration examples for classic and modern modes

tanrax committed:
Fix classic mode navigation

Restored original button-based navigation for classic UI.

tanrax commented:
I want to mention that I'm not particularly happy with maintaining both the current and new appearances. The code has become more complex. It's been very difficult to keep all the functionalities of both the current and new code. Perhaps the problem is architectural; the logic isn't properly separated from the UI

tanrax committed:
Remove Lobsters references

tanrax committed:
Fix buffer state on load more stories

Move hackernews-mode activation before setting local variables.
Remove redundant kill-all-local-variables and read-only-mode calls.

tanrax committed:
Filter null items before rendering

Remove deleted/dead stories that return null from API.

tanrax committed:
Fix keymap hierarchy and filter dead items

Use composed keymap for modern UI to preserve all keybindings.
Filter deleted and dead items in addition to null.

tanrax committed:
Add format parameter to comments button widget

basil-conto commented:
Thanks for continuing to work on this! And sorry for the very late reply.

Sadly I'm still a bit too tied up to review this properly, but I hope to get a chance soon.
I'll also have a look at rearchitecting the package a bit to make the feature easier to integrate.

On that note, just as a reminder to look at later, I recently heard of another promising UI package which has been submitted for inclusion in GNU ELPA: https://lists.gnu.org/r/emacs-devel/2026-01/msg00470.html

tanrax commented:
@basil-conto Don't worry, there is no rush.

Regarding the architecture, if we follow a simple MVC pattern, we can have different Views/UIs. I don't think it will be difficult to achieve. I am open to deleting my work and starting from scratch if that makes the transformation easier.

On the other hand, the new UI framework, in my opinion, we don't need it. Just using widgets, which are already integrated into the core, we can achieve very good results. Check a screenshot of the current code.

<img width="890" height="908" alt="image" src="https://github.com/user-attachments/assets/88941acf-0155-415a-a582-9f3f7ddaaaff" />
